File: Rust/2020/01.rs
#![feature(test)]

use std::collections::HashSet;

type Input = Vec<i32>;

fn setup(input: &str) -> Input {
    input.lines().map(|i| i.parse().unwrap()).collect()
}

fn part1(input: &Input) -> i32 {
    let mut seen: HashSet<i32> = HashSet::new();
    for &x in input {
        let y = 2020 - x;
        if seen.contains(&y) {
            return x * y;
        }
        seen.insert(x);
    }
    panic!();
}

fn part2(input: &Input) -> i32 {
    for (i, &x) in input.iter().enumerate() {
        let mut seen: HashSet<i32> = HashSet::new();
        for &y in &input[i + 1..] {
            let z = 2020 - x - y;
            if seen.contains(&z) {
                return x * y * z;
            }
            seen.insert(y);
        }
    }
    panic!();
}

aoc::main!(2020, 1, ex: 1);

File: Python/2020/01.py
from lib import *

input = read_input(2020, 1)

nums = ints(input)


seen = set()
for a in nums:
    b = 2020 - a
    if b in seen:
        print(a * b)
        break
    seen.add(a)


for i, a in enumerate(nums):
    seen = set()
    for b in nums[i + 1 :]:
        c = 2020 - a - b
        if c in seen:
            print(a * b * c)
            break
        seen.add(b)
    else:
        continue
    break
